File: basic/ts/src/basicspec.test.ts
// The TS half of the base-layer parity corpus (basic/spec/*.tsv).
//
// basic/go/basicspec_test.go is the other half. The two runners share NO
// code — they read the same files and each builds the registry and renders
// the residual independently, exactly as core/spec's and parser/spec's
// pairs do. That independence is the point: shared scaffolding can hide
// the same bug from both engines (design/CORE-GO-TS-DEFECTS.0.md, blind
// spot 9).
//
// This is a SPEC, not a differential: the `expected` column is the
// documented contract, so a row can legitimately fail on BOTH engines.

import { describe, it } from 'node:test'
import { strict as assert } from 'node:assert'
import * as fs from 'node:fs'
import * as path from 'node:path'
import { fileURLToPath } from 'node:url'

import {
  BoruError,
  canon,
  Engine,
  newInteger,
  newString,
  newWord,
  Registry,
  type Value,
} from '@boru-lang/core'
import { stackNatives } from './native-stack.ts'

const __dirname = path.dirname(fileURLToPath(import.meta.url))
const SPEC_DIR = path.resolve(__dirname, '..', '..', 'spec')

interface SpecRow {
  file: string
  line: number
  prog: string
  expected: string
  note: string
}

function readSpec(file: string): SpecRow[] {
  const text = fs.readFileSync(path.join(SPEC_DIR, file), 'utf8')
  const rows: SpecRow[] = []
  const lines = text.split('\n')
  for (let n = 0; n < lines.length; n++) {
    const line = lines[n]!
    // A line is a COMMENT only when it starts with '#' AND carries no tab —
    // '#' is boru's own comment marker, so a program may begin with one.
    if ('' === line.trim() || (line.startsWith('#') && !line.includes('\t'))) continue
    const parts = line.split('\t')
    assert.ok(parts.length >= 2, `${file}:${n + 1}: need at least 2 tab-separated columns`)
    rows.push({
      file,
      line: n + 1,
      prog: parts[0]!,
      expected: parts[1]!,
      note: parts[2] ?? '',
    })
  }
  return rows
}

/**
 * token builds one program token: a decimal is an Integer, '…' is a
 * String, anything else is a Word. Deliberately tiny — the corpus is about
 * the WORDS, so the token notation stays out of the way.
 */
function token(tok: string): Value {
  if (tok.startsWith("'") && tok.endsWith("'") && tok.length >= 2) {
    return newString(tok.slice(1, -1))
  }
  if (/^-?\d+$/.test(tok)) return newInteger(BigInt(tok))
  return newWord(tok)
}

/** Registers ONLY the stack vocabulary, so a row depends on nothing else. */
function specRegistry(): Registry {
  const r = new Registry()
  for (const nf of stackNatives) r.registerNativeFunc(nf)
  return r
}

function runProgram(prog: string): string {
  const toks = prog.split(/\s+/).filter((s) => '' !== s).map(token)
  try {
    return canon(new Engine(specRegistry()).run(toks))
  } catch (e) {
    if (e instanceof BoruError) return `ERROR:${e.code}`
    return `ERROR:non_boru:${(e as Error).message}`
  }
}

describe('basic spec — stack.tsv', () => {
  const rows = readSpec('stack.tsv')
  assert.ok(rows.length > 0, 'basic/spec produced no rows — the corpus is not being read')
  for (const r of rows) {
    it(`${r.line}: ${r.prog}`, () => {
      assert.equal(runProgram(r.prog), r.expected, r.note)
    })
  }
})

File: basic/go/basicspec_test.go
package basic_test

// The Go half of the base-layer parity corpus (basic/spec/*.tsv).
//
// basic/ts/src/basicspec.test.ts is the other half. The two runners share
// NO code — they read the same files and each builds the registry and
// renders the residual independently, exactly as core/spec's and
// parser/spec's pairs do. That independence is the point: shared
// scaffolding can hide the same bug from both engines
// (design/CORE-GO-TS-DEFECTS.0.md, blind spot 9).
//
// This is a SPEC, not a differential: the `expected` column is the
// documented contract, so a row can legitimately fail on BOTH engines —
// which is exactly the class of defect an agreement-only corpus is
// structurally blind to.

import (
	"bufio"
	"errors"
	"os"
	"path/filepath"
	"strconv"
	"strings"
	"testing"

	basic "github.com/boru-lang/boru/basic/go"
	core "github.com/boru-lang/boru/core/go"
)

const basicSpecDir = "../spec"

type basicSpecRow struct {
	file     string
	line     int
	prog     string
	expected string
	note     string
}

func parseBasicSpec(t *testing.T, path string) []basicSpecRow {
	t.Helper()
	f, err := os.Open(path)
	if err != nil {
		t.Fatalf("open %s: %v", path, err)
	}
	defer func() { _ = f.Close() }()

	var rows []basicSpecRow
	sc := bufio.NewScanner(f)
	sc.Buffer(make([]byte, 1<<20), 1<<20)
	for n := 1; sc.Scan(); n++ {
		line := sc.Text()
		// A line is a COMMENT only when it starts with '#' AND carries no
		// tab — '#' is boru's own comment marker, so a program may begin
		// with one.
		if strings.TrimSpace(line) == "" ||
			(strings.HasPrefix(line, "#") && !strings.Contains(line, "\t")) {
			continue
		}
		parts := strings.Split(line, "\t")
		if len(parts) < 2 {
			t.Fatalf("%s:%d: want at least 2 tab-separated columns, got %d", path, n, len(parts))
		}
		row := basicSpecRow{file: filepath.Base(path), line: n, prog: parts[0], expected: parts[1]}
		if len(parts) > 2 {
			row.note = parts[2]
		}
		rows = append(rows, row)
	}
	if err := sc.Err(); err != nil {
		t.Fatalf("read %s: %v", path, err)
	}
	return rows
}

// basicSpecToken builds one program token: a decimal is an Integer, '…' is
// a String, anything else is a Word. Deliberately tiny — the corpus is
// about the WORDS, so the token notation stays out of the way.
func basicSpecToken(tok string) core.Value {
	if strings.HasPrefix(tok, "'") && strings.HasSuffix(tok, "'") && len(tok) >= 2 {
		return core.NewString(tok[1 : len(tok)-1])
	}
	if n, err := strconv.ParseInt(tok, 10, 64); err == nil {
		return core.NewInteger(n)
	}
	return core.NewWord(tok)
}

// basicSpecRegistry registers ONLY the stack vocabulary, so a row's result
// depends on nothing else in the layer.
func basicSpecRegistry(t *testing.T) *core.Registry {
	t.Helper()
	r, err := core.NewRegistry()
	if err != nil {
		t.Fatalf("NewRegistry: %v", err)
	}
	r.InitRootContext()
	for _, nf := range basic.StackNatives {
		r.RegisterNativeFunc(nf)
	}
	if err := r.Err(); err != nil {
		t.Fatalf("registration: %v", err)
	}
	return r
}

func TestBasicSpec(t *testing.T) {
	entries, err := os.ReadDir(basicSpecDir)
	if err != nil {
		t.Fatalf("read %s: %v", basicSpecDir, err)
	}
	total := 0
	for _, e := range entries {
		if !strings.HasSuffix(e.Name(), ".tsv") {
			continue
		}
		path := filepath.Join(basicSpecDir, e.Name())
		rows := parseBasicSpec(t, path)
		t.Run(strings.TrimSuffix(e.Name(), ".tsv"), func(t *testing.T) {
			for _, row := range rows {
				total++
				var prog []core.Value
				for _, tok := range strings.Fields(row.prog) {
					prog = append(prog, basicSpecToken(tok))
				}
				got := ""
				out, rerr := core.NewTop(basicSpecRegistry(t)).Run(prog)
				if rerr != nil {
					var be *core.BoruError
					if errors.As(rerr, &be) {
						got = "ERROR:" + be.Code
					} else {
						got = "ERROR:non_boru:" + rerr.Error()
					}
				} else {
					parts := make([]string, 0, len(out))
					for _, v := range out {
						parts = append(parts, core.CanonValue(v))
					}
					got = strings.Join(parts, " ")
				}
				if got != row.expected {
					t.Errorf("%s:%d  %s\n  got  %q\n  want %q\n  (%s)",
						row.file, row.line, row.prog, got, row.expected, row.note)
				}
			}
		})
	}
	if total == 0 {
		t.Fatal("basic/spec produced no rows — the corpus is not being read")
	}
	t.Logf("basic/spec: %d rows", total)
}
